File: pset6/sentimental-credit/credit.py
import sys


def main():
    credit_card = get_card_num()

    validate_card(credit_card)
    

def get_card_num():
    while True:
        cardnumber = input("Number: ")
        try:
            if len(cardnumber) > 0 and int(cardnumber):
                break
        except ValueError:
            continue
    return cardnumber


def validate_card(credit_card):
    digits = [13, 15, 16]
    card_len = len(credit_card)
    total = div = mod = 0
    left = []
    underline = []

    # print(card_len)

    if card_len not in digits:
        print("INVALID")
        sys.exit(0)

    if (card_len % 2 == 0):
        for i in range(0, card_len, 2):
            if int(credit_card[i]) * 2 < 10:
                underline.append(int(credit_card[i]) * 2)
            else:
                div = int(credit_card[i]) * 2 // 10
                mod = int(credit_card[i]) * 2 % 10
                underline.append(div)
                underline.append(mod)
        # print("underline: " + str(underline))
        for j in range(1, card_len, 2):
            left.append(int(credit_card[j]))
        # print("left digits: " + str(left))
        total = int(sum(underline)) + int(sum(left))
        # print("even digits")
        # print("total: " + str(total))
    else:
        for i in range(1, card_len, 2):
            if int(credit_card[i]) * 2 < 10:
                underline.append(int(credit_card[i]) * 2)
            else:
                div = int(credit_card[i]) * 2 // 10
                mod = int(credit_card[i]) * 2 % 10
                underline.append(div)
                underline.append(mod)
        for j in range(0, card_len, 2):
            left.append(int(credit_card[j]))
        total = int(sum(underline)) + int(sum(left))
        # print("odd digits")
        # print(total)

    if (total % 10 != 0):
        print("INVALID")

        sys.exit(0)

    # check card type
    # First two digits
    first = int(credit_card[0])
    second = int(credit_card[1])

    if first == 3 and (second == 4 or second == 7) and card_len == 15:
        print("AMEX")
    elif first == 5 and (second <= 5 and second > 0) and card_len == 16:
        print("MASTERCARD")
    elif first == 4 and (card_len == 13 or card_len == 16):
        print("VISA")
    else:
        print("INVALID")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long n = get_long("Number: ");


// check if the card is valid by implementing Luhn's algorithm

    int sum_1 = 0;
    int sum_2 = 0;
    int sum = 0;
    int i = 1;

    long cardnumber = n;
    do
    {
        int digit = cardnumber % 10; 
        if (i % 2 != 0)
        {
            sum_1 = sum_1 + digit;
            //printf("odd digit: %i\n",digit);
        }
        else
        {
            int product = digit * 2;
            if (product < 10)
            {
                sum_2 = sum_2 + product;
            }
            else
            {
                sum_2 = sum_2 + product / 10 + product % 10;
            }
            //printf("even digit: %i\n",digit);
        }
        sum = sum_1 + sum_2;
        i = i + 1;
        cardnumber = cardnumber / 10;

    }
    while (cardnumber > 0);

    //printf("sum_1: %i\n",sum_1);
    //printf("sum_2: %i\n",sum_2);

    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        //printf("The sum is %i\n",sum);
        return 0;
    }

// check card type

    // First two digits
    long start_2 = n;
    int j = i - 1;
    if (j % 2 == 0)
    {
        do
        {
            start_2 = start_2 / 100;
            //printf("The quotient: %li\n", start_2);
        }
        while (start_2 > 100);
        //printf("The first two digits: %li\n", start_2);
    }
    else
    {
        start_2 = n / 10;
        do
        {
            start_2 = start_2 / 100;
        }
        while (start_2 > 100);
        //printf("The first two digits: %li\n", start_2);
    }

    // First digit
    long start_1 = n;
    while (start_1 > 10)
    {
        start_1 = start_1 / 10;
    }
    //printf("The first digit: %li\n", start_1);

    if (j == 15 && (start_2 == 34 || start_2 == 37))
    {
        printf("AMEX\n");
    }

    else if (j == 16 && (start_2 > 50 && start_2 < 56))
    {
        printf("MASTERCARD\n");
    }

    else if ((j == 16 || j == 13) && start_1 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}
